File: Native/Android/LunarConsole/lunarConsole/src/main/java/spacemadness/com/lunarconsole/rx/BehaviorSubject.java
package spacemadness.com.lunarconsole.rx;

import spacemadness.com.lunarconsole.core.Disposable;

public final class BehaviorSubject<T> extends Subject<T> {
    private T value;

    public BehaviorSubject() {
        this(null);
    }

    public BehaviorSubject(T value) {
        this.value = value;
    }

    @Override
    public Disposable subscribe(Observer<T> observer) {
        Disposable disposable = super.subscribe(observer);
        notifyObserver(observer, getValue());
        return disposable;
    }

    public T getValue() {
        return value;
    }

    public void setValue(T value) {
        this.value = value;
        notifyObservers(value);
    }
}

File: Native/Android/LunarConsoleNew/lunarCore/src/main/java/spacemadness/com/lunarconsole/reactive/BehaviorSubject.kt
package spacemadness.com.lunarconsole.reactive

import spacemadness.com.lunarconsole.core.Disposable

class BehaviorSubject<T>(value: T) : Subject<T>() {
    var value: T = value
        set(value) {
            field = value
            notifyObservers(value)
        }

    override fun subscribe(observer: (T) -> Unit): Disposable {
        val observe = super.subscribe(observer)
        notifyObserver(observer, value)
        return observe
    }
}